### PR TITLE
Features/compiler tests

### PR DESCRIPTION
**Pull request by becker33**

### CODE REVIEW

File: var/spack/tests/compilers/cxx/hello.c++ [hunk 1]
```diff
@@ -0,0 +1,10 @@
+#include <stdio.h>
+
+int main()
+{
+
+	printf ("Hello world from C++\n");
```
Comment (tldahlgren) [inline var/spack/tests/compilers/cxx/hello.c++]: Should `printf` be used here for a `.c++` file?

File: var/spack/tests/compilers/fortran/hello.f95 [hunk 2]
```diff
@@ -0,0 +1,6 @@
+      program line
+
+      write (*,*) "Hello world from FORTRAN"
+      write (*,*) "YES!"
+
+      end
```
Comment (tldahlgren) [inline var/spack/tests/compilers/fortran/hello.f95]: Could get rid of the indentation before `program` and `end` and terminate with `end program line`.

**Review by tldahlgren — COMMENTED** [review]
Would be more "interesting" to exercise a few more features of the different versions of the languages to ensure they are supported by the compilers.

For example, the FORTRAN examples can reflect indentation differences (e.g., F77 requires indentation, F90+ do not) and string variable declarations (e.g. different ways for declaring a string to hold the name of the language to be output in the print/write statement).

File: var/spack/tests/compilers/cxx/hello.c++ (hunk 1 above)
Comment (becker33) [reply var/spack/tests/compilers/cxx/hello.c++]: There are multipe C++ tests. C++ compilers need to be able to handle C-style prints.

**Comment by becker33** [comment]
@tldahlgren I agree that there's a lot more we could (and should eventually) test. The first thing that comes to mind is the different standard versions. I did incorporate your suggestion for f95 to test f90 formatting features. But the c++ 14 and later standards would be the next thing I would test.

For right now though, I'm more concerned about getting some basic tests in place that can be the basis for future work, not trying to test every feature of each compiler.

File: lib/spack/spack/build_environment.py [hunk 3]
```diff
@@ -728,25 +728,25 @@ def setup_package(pkg, dirty, context='build'):
...
+        if (not dirty) and (not env.is_unset('CPATH')):
+            tty.debug("A dependency has updated CPATH, this may lead pkg-config"
+                      " to assume that the package is part of the system"
+                      " includes and omit it when invoked with '--cflags'.")
+
+        # setup package itself
+        set_module_variables_for_package(pkg)
         pkg.setup_build_environment(env)
     elif context == 'test':
+        # one command encapsulates all of the above for run environments
```
Comment (tgamblin) [inline lib/spack/spack/build_environment.py]: Not sure what this comment means -- does the command below do all the same stuff?  Seems like it doesn't.

Shouldn't the test environment make use of the same module-scope variables as the build environment?  It's frequently caused issues in the past for certain functions (like `patch()` and stuff that runs before `install()`) when these variables have been missing -- e.g., packages have assumed `spack_cc` but it hasn't been there for certain commands.

File: lib/spack/spack/build_systems/intel.py [hunk 4]
```diff
@@ -1017,6 +1017,15 @@ def setup_run_environment(self, env):
 
         env.extend(EnvironmentModifications.from_sourcing_file(f, *args))
 
+        if self.spec.name in ('intel', 'intel-parallel-studio'):
+            # this package provides compilers
+            # TODO: fix check above when compilers are dependencies
```
Comment (tgamblin) [inline lib/spack/spack/build_systems/intel.py:1022]: I am worried that this can go wrong, particularly in environments.  See this issue: https://github.com/spack/spack/issues/17823#issuecomment-667448964.

`setup_run_environment` is called recursively for all dependencies in the environment, but we don't know whether the user intends to build something with each and every package.  In the issue above, @ax3l set `CUDAHOSTCXX`, but users were trying to load the environment and build with *other* compilers than the one CUDA was built with.  I think setting `CC` et al. here puts us in the same position.  What if `gcc` sets these too?  You could get a mixed-compiler environment where there is a race for which dependency gets to set `CC` last, which sounds horrible.

I think we need a way to say that a run environment is to be used for building or testing, and to say what tools we want to build with -- then only those tools should be setting these types of behavior-affecting variables.  Is there an easy way to do something like that here?

File: lib/spack/spack/build_systems/intel.py [hunk 5]
```diff
@@ -1309,3 +1318,48 @@ def uninstall_ism(self):
...
+
+                # Get the right compiler for this test
+                compiler = test_compiler_for_source(test)
+
+                ###
+                # Compile the test
+                ###
+                # once we reconcile compilers and dependencies this should
+                # query package for appropriate cxx11 flag
+                # only the hello_cxx11 test requires c++11
```
Comment (tgamblin) [inline lib/spack/spack/build_systems/intel.py:1347]: Is the idea that this will move to a virtual package suitable for any compiler, so we'll need to query?

File: lib/spack/spack/cmd/clean.py [hunk 6]
```diff
@@ -9,6 +9,7 @@
 import collections
 
 import llnl.util.tty as tty
+import llnl.util.filesystem as fs
```
Comment (tgamblin) [inline lib/spack/spack/cmd/clean.py]: This seems to be unused.  Was it supposed to be imported in the test for `clean` (which uses `fs`)?

File: lib/spack/spack/cmd/list.py [hunk 7]
```diff
@@ -54,6 +54,9 @@ def setup_parser(subparser):
     subparser.add_argument(
         '--update', metavar='FILE', default=None, action='store',
         help='write output to the specified file, if any package is newer')
+    subparser.add_argument(
+        '-v', '--include-virtuals', action='store_true', default=False,
```
Comment (tgamblin) [inline lib/spack/spack/cmd/list.py]: `--virtuals` (to be consistent with other places like `find` help).

Also need to run `spack commaands --update-completion`

File: lib/spack/spack/package.py [hunk 8]
```diff
@@ -1558,25 +1563,59 @@ def test_process():
                     shutil.rmtree(testdir)
                 fsys.mkdirp(testdir)
 
-                # copy test data into testdir/data
-                datadir = Prefix(self.package_dir).test
-                if os.path.isdir(datadir):
-                    shutil.copytree(datadir, testdir.data)
+                # cd to test directory
+                os.chdir(testdir)
```
Comment (tgamblin) [inline lib/spack/spack/package.py]: any chdir like this should be wrapped in try/finally via `with working_dir()`

File: lib/spack/spack/package.py [hunk 9]
```diff
@@ -1558,25 +1563,59 @@ def test_process():
...
+
+                # run test methods from the package and all virtuals it provides
+                # virtuals have to be deduped by name
+                v_names = list(set([vspec.name
+                                    for vspec in self.virtuals_provided]))
```
Comment (tgamblin) [inline lib/spack/spack/package.py:1569]: Minor: The order of `v_names` (and as a result also `test_specs`) is nondeterministic.  We should make the final list `sorted` so that the tests are done in the same order each time -- just easier for the user to make a change and run the command again and see that nothing else changed.

File: lib/spack/spack/package.py [hunk 10]
```diff
@@ -1558,25 +1563,59 @@ def test_process():
...
-                        if not os.listdir(self.test_stage):
-                            shutil.rmtree(self.test_stage)
-                finally:
-                    # reset debug level
-                    tty.set_debug(old_debug)
+                    # copy test data into testdir/data
+                    data_source = Prefix(spec_pkg.package_dir).test
+                    data_dir = os.path.join(testdir.data, spec.name)
+                    if os.path.isdir(data_source):
+                        shutil.copytree(data_source, data_dir)
```
Comment (tgamblin) [inline lib/spack/spack/package.py]: are tests always owned by the person running them?  Any permissions here?

File: lib/spack/spack/package.py [hunk 11]
```diff
@@ -1558,25 +1563,59 @@ def test_process():
...
+
+                    try:
+                        # grab the function for each method so we can all it
```
Comment (tgamblin) [inline lib/spack/spack/package.py]: call

File: lib/spack/spack/package.py [hunk 12]
```diff
@@ -1558,25 +1563,59 @@ def test_process():
...
+                        # with this package in place of its `self` object
+                        test_fn = spec_pkg.__class__.test
+                        if not isinstance(test_fn, types.FunctionType):
+                            test_fn = test_method.__func__
+                        test_fn(self)
+                    except BaseException:
+                        # reset debug level on failfast errors
+                        tty.set_debug(old_debug)
```
Comment (tgamblin) [inline lib/spack/spack/package.py]: should this not be in a `finally` block?

File: lib/spack/spack/package.py [hunk 13]
```diff
@@ -1558,25 +1563,59 @@ def test_process():
...
+                        raise
+
+                if self.test_failures:
+                    raise TestFailure(self.test_failures)
+
+                # cleanup test directory on success
+                if remove_directory:
+                    shutil.rmtree(testdir)
+                    if not os.listdir(self.test_stage):
```
Comment (tgamblin) [inline lib/spack/spack/package.py]: what's this testing?  if it's empty we leave it, but if it's not empty we completely remove it?  Seems like the final state should be empty or no dir either way -- pick one

File: lib/spack/spack/package.py [hunk 14]
```diff
@@ -1558,25 +1563,59 @@ def test_process():
...
+                        shutil.rmtree(self.test_stage)
+
+                # reset debug level
+                tty.set_debug(old_debug)
```
Comment (tgamblin) [inline lib/spack/spack/package.py]: I think this belongs in a `finally` block around the whole thing.  It will currently be skipped if there's any filesystem error in the cleanup code.  Looks like it was in a `finally` originally -- why the change?

File: lib/spack/spack/repo.py [hunk 15]
```diff
@@ -992,9 +995,12 @@ def _pkg_checker(self):
...
         return self._fast_package_checker
 
-    def all_package_names(self):
+    def all_package_names(self, include_virtuals=False):
         """Returns a sorted list of all package names in the Repo."""
-        return sorted(self._pkg_checker.keys())
+        names = sorted(self._pkg_checker.keys())
+        if include_virtuals:
+            return names
+        return list(filter(lambda x: not self.is_virtual(x), names))
```
Comment (tgamblin) [inline lib/spack/spack/repo.py]: use a list comprehension

File: lib/spack/spack/spec.py [hunk 16]
```diff
@@ -1196,7 +1196,9 @@ def virtual(self):
     @staticmethod
     def is_virtual(name):
         """Test if a name is virtual without requiring a Spec."""
-        return (name is not None) and (not spack.repo.path.exists(name))
+        return (name is not None and
+                (not spack.repo.path.exists(name) or
+                 spack.repo.path.get_pkg_class(name).virtual))
```
Comment (tgamblin) [inline lib/spack/spack/spec.py]: It's not clear who is ultimately responsible for determining if a package is virtual.  I think this logic needs to get folded into the repo -- right now this consults the repo *and* the class, but there should really be one source of truth.  Can you fold this logic into `is_virtual`?  I realize that does the caching, so it may take some thought to figure out how to structure this.

File: lib/spack/spack/test/cmd/clean.py [hunk 17]
```diff
@@ -38,14 +38,14 @@ def __call__(self, *args, **kwargs):
...
-    ('-a',       [0, 1, 1, 1, 1]),
-    ('',         [0, 0, 0, 0, 0]),
+    ('mpileaks', [1, 0, 0, 0, 0, 0]),
+    ('-s',       [0, 1, 0, 0, 0, 0]),
+    ('-sd',      [0, 1, 1, 0, 0, 0]),
+    ('-m',       [0, 0, 0, 1, 0, 0]),
+    ('-f',       [0, 0, 0, 0, 1, 0]),
+    ('-t',       [0, 0, 0, 0, 0, 1]),
+    ('-a',       [0, 1, 1, 1, 1, 1]),
+    ('',         [0, 0, 0, 0, 0, 0]),
```
Comment (tgamblin) [inline lib/spack/spack/test/cmd/clean.py]: counters should really be a dict with informatively named keys.

File: lib/spack/spack/util/executable.py [hunk 18]
```diff
@@ -146,7 +146,7 @@ def streamify(arg, mode):
 
         quoted_args = [arg for arg in args if re.search(r'^"|^\'|"$|\'$', arg)]
         if quoted_args:
-            tty.warn(
+            tty.debug(
```
Comment (tgamblin) [inline lib/spack/spack/util/executable.py]: Are there times this is just a nuisance and not a real bug?  I think it's actually usually worrisome to see this.

File: var/spack/repos/builtin/packages/c/package.py [hunk 19]
```diff
@@ -0,0 +1,26 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import os
+
+class C(Package):
+    """Virtual package for the C language."""
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/c/package.py]: might want to say " for C compilers"

File: var/spack/repos/builtin/packages/c/test/hello.c [hunk 20]
```diff
@@ -0,0 +1,7 @@
+#include <stdio.h>
+int main()
+{
+	printf ("Hello world from C!\n");
+	printf ("YES!");
+	return (0);
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/c/test/hello.c]: why `(0)` and not 0?

File: var/spack/repos/builtin/packages/c/test/hello.c [hunk 21]
```diff
@@ -0,0 +1,7 @@
+#include <stdio.h>
+int main()
+{
+	printf ("Hello world from C!\n");
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/c/test/hello.c:4]: Style:

```
    printf("Hello world from C!\n");
```

Comment (tgamblin) [reply var/spack/repos/builtin/packages/c/test/hello.c:4]: I guess if we are getting technical about this, this exists: https://www.python.org/dev/peps/pep-0007/

File: var/spack/repos/builtin/packages/cxx/test/hello.c++ [hunk 22]
```diff
@@ -0,0 +1,10 @@
+#include <stdio.h>
+
+int main()
+{
+
+	printf ("Hello world from C++\n");
+	printf ("YES!");
+
+	return (0);
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/cxx/test/hello.c++]: why (0)

File: var/spack/repos/builtin/packages/cxx/test/hello.c++ [hunk 23]
```diff
@@ -0,0 +1,10 @@
+#include <stdio.h>
+
+int main()
+{
+
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/cxx/test/hello.c++]: why blank here and not above?

File: var/spack/repos/builtin/packages/cxx/test/hello.cc [hunk 24]
```diff
@@ -0,0 +1,10 @@
+#include <iostream>
+using namespace std;
+
+int main()
+{
+
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/cxx/test/hello.cc]: why blank?

File: var/spack/repos/builtin/packages/cxx/test/hello.cc [hunk 25]
```diff
@@ -0,0 +1,10 @@
...
+       cout << "Hello world from C++!" << endl;
+       cout << "YES!" << endl;
+	return (0);
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/cxx/test/hello.cc:8]: why (0)?

File: var/spack/repos/builtin/packages/cxx/test/hello.cc [hunk 26]
```diff
@@ -0,0 +1,10 @@
+#include <iostream>
+using namespace std;
+
+int main()
+{
+
+       cout << "Hello world from C++!" << endl;
+       cout << "YES!" << endl;
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/cxx/test/hello.cc]: why is this indented differently from return?

File: var/spack/repos/builtin/packages/cxx/test/hello.cpp [hunk 27]
```diff
@@ -0,0 +1,10 @@
+#include <iostream>
+using namespace std;
+
+int main()
+{
+
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/cxx/test/hello.cpp]: no blank

File: var/spack/repos/builtin/packages/cxx/test/hello.cpp [hunk 28]
```diff
@@ -0,0 +1,10 @@
...
+       cout << "Hello world from C++!" << endl;
+       cout << "YES!" << endl;
+	return (0);
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/cxx/test/hello.cpp]: indentation, (0) -> 0

File: var/spack/repos/builtin/packages/fortran/test/hello.f90 [hunk 29]
```diff
@@ -0,0 +1,6 @@
+program line
+
+      write (*,*) "Hello world from FORTRAN"
+      write (*,*) "YES!"
+
+end program line
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/fortran/test/hello.f90:6]: might be nice to have f03

File: var/spack/repos/builtin/packages/gcc/package.py [hunk 30]
```diff
@@ -11,6 +11,7 @@
 import itertools
 import os
 import sys
+import shutil
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/gcc/package.py]: not used?

File: var/spack/repos/builtin/packages/intel/package.py [hunk 31]
```diff
@@ -3,6 +3,8 @@
 #
 # SPDX-License-Identifier: (Apache-2.0 OR MIT)
 
+import shutil
+import os
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/intel/package.py]: not used?

File: var/spack/repos/builtin/packages/llvm/package.py [hunk 32]
```diff
@@ -5,6 +5,8 @@
 
 from spack import *
 import sys
+import shutil
+import os
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/llvm/package.py]: not used?

File: var/spack/repos/builtin/packages/mpi/test/mpi_hello.c [hunk 33]
```diff
@@ -0,0 +1,15 @@
+#include <mpi.h>
+#include <stdio.h>
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/mpi/test/mpi_hello.c:1]: move system headers before MPI (a library)

File: var/spack/repos/builtin/packages/mpi/test/mpi_hello.c [hunk 34]
```diff
@@ -0,0 +1,15 @@
...
+
+int main(int argc, char** argv){
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/mpi/test/mpi_hello.c]: `) {`

File: var/spack/repos/builtin/packages/mpi/test/mpi_hello.c [hunk 35]
```diff
@@ -0,0 +1,15 @@
...
+  MPI_Init(argc, argv);
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/mpi/test/mpi_hello.c]: &argc, &argv

File: var/spack/repos/builtin/packages/mpi/test/mpi_hello.c [hunk 36]
```diff
@@ -0,0 +1,15 @@
...
+
+  int rank;
+  int num_ranks;
+  MPI_Comm_size(MPI_COMM_WORLD, &num_ranks);
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
```
Comment (tgamblin) [inline var/spack/repos/builtin/packages/mpi/test/mpi_hello.c:9]: make this order match decl order (it's weirdly reversed)

**Review by tgamblin — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/llvm/package.py (hunk 32 above)
Comment (becker33) [reply var/spack/repos/builtin/packages/llvm/package.py]: Yeah, used in deleted code. I'll pull those out.

**Comment by tldahlgren** [comment]
> For right now though, I'm more concerned about getting some basic tests in place that can be the basis for future work, not trying to test every feature of each compiler.

Of course.  These are "smoke" tests.  I was not suggesting every feature of each compiler be tested.

File: lib/spack/spack/package.py (hunk 13 above)
Comment (becker33) [reply lib/spack/spack/package.py]: I think you read this backwards: It it's empty we remove it, if it has something in it we leave it.

Comment (tgamblin) [reply lib/spack/spack/package.py]: yep you're right

File: lib/spack/spack/cmd/list.py (hunk 7 above)
Comment (becker33) [reply lib/spack/spack/cmd/list.py]: There isn't a `--virtuals` option anywhere that I can find to be consistent with, but I can still make the change.

File: lib/spack/spack/package.py (hunk 10 above)
Comment (becker33) [reply lib/spack/spack/package.py]: There shouldn't be any permissions issue here. The tests are being copied from the spack package dir (which shouldn't have permissions issues), and the copies will be owned by the person running the tests

File: var/spack/repos/builtin/packages/c/test/hello.c (hunk 21 above)
Comment (becker33) [reply var/spack/repos/builtin/packages/c/test/hello.c:4]: I don't think we should use a style guide on the code we're using to test the compilers. There is value in throwing unusual combinations of things at a compiler (although we obviously are not testing comprehensively enough to do so in a rigorous manner).

Comment (becker33) [reply var/spack/repos/builtin/packages/c/test/hello.c:4]: These tests come from John G. I'm assuming that most of the weirdness in the test code is an accumulation of individual errors that some compiler has had over the course of his career.

File: var/spack/repos/builtin/packages/cxx/test/hello.c++ (hunk 22 above)
Comment (becker33) [reply var/spack/repos/builtin/packages/cxx/test/hello.c++]: I'm assuming it's there because some compiler failed on that in the past

File: lib/spack/spack/package.py (hunk 12 above)
Comment (becker33) [reply lib/spack/spack/package.py]: The `try` statement is inside a `for`-loop, because it iterates over the package and all virtuals it provides. We need to run this any time we're exiting the loop, but sometimes we need to exit the loop early. I might be able to engineer it into a finally block by layering the `try` statements though, I'll take a look at it.

File: lib/spack/spack/package.py (hunk 14 above)
Comment (becker33) [reply lib/spack/spack/package.py]: See above

File: lib/spack/spack/package.py (hunk 8 above)
Comment (becker33) [reply lib/spack/spack/package.py]: I didn't think that was necessary since this is executing in a subprocess, but I can change it.

File: lib/spack/spack/cmd/clean.py (hunk 6 above)
Comment (becker33) [reply lib/spack/spack/cmd/clean.py]: It was for older code that was removed.

File: lib/spack/spack/package.py (hunk 12 above)
Comment (becker33) [reply lib/spack/spack/package.py]: Fixed.

File: lib/spack/spack/build_systems/intel.py (hunk 5 above)
Comment (becker33) [reply lib/spack/spack/build_systems/intel.py:1347]: Eventually it would be part of the contract for the `cxx` virtual package.

Before we implement contracts for virtuals, it would be defined individually on the compiler packages, the way it's currently defined in `lib/spack/spack/compilers/*.py`

File: lib/spack/spack/cmd/list.py (hunk 7 above)
Comment (tgamblin) [reply lib/spack/spack/cmd/list.py]: I meant more that the "show X" options there are named "--X"

File: lib/spack/spack/build_environment.py (hunk 3 above)
Comment (becker33) [reply lib/spack/spack/build_environment.py]: I'm worried about having variables in scope that aren't usable (say, the compiler for tests that don't require the compiler). For now I'll add it.

File: lib/spack/spack/package.py (hunk 10 above)
Comment (tgamblin) [reply lib/spack/spack/package.py]: 👍

File: lib/spack/spack/build_systems/intel.py (hunk 4 above)
Comment (becker33) [reply lib/spack/spack/build_systems/intel.py:1022]: I have a TODO note in the code to change this once we have compilers as dependencies.

I think the tricky bit is `intel-parallel-studio`. I can't come up with a reason we would have an `intel` dep without wanting the compilers in the run environment (and in general, modules for the intel compiler add these variables).

For `intel-parallel-studio`, I think we need to check the edge information when accessing it as a dependency to know whether it's providing c/cxx/fortran in this particular instance. But we don't have the capability to do that yet.

For environments, we probably need to have deptypes eventually, but that's a whole 'nother feature.

File: lib/spack/spack/util/executable.py (hunk 18 above)
Comment (becker33) [reply lib/spack/spack/util/executable.py]: I thought this was coming up repeatedly on non-problematic code, but I must have edited that away. I'll revert this.

File: lib/spack/spack/spec.py (hunk 16 above)
Comment (becker33) [reply lib/spack/spack/spec.py]: It definitely did take some work, but it proved doable. I had to add a parameter for whether to allow use of the index for a particular call to `repo.path.is_virtual`.

File: lib/spack/spack/test/cmd/clean.py (hunk 17 above)
Comment (becker33) [reply lib/spack/spack/test/cmd/clean.py]: ~I agree, but it's non-trivial to refactor and I think it's outside the scope of this PR.~

But then I did it anyway

File: var/spack/repos/builtin/packages/fortran/test/hello.f90 (hunk 29 above)
Comment (becker33) [reply var/spack/repos/builtin/packages/fortran/test/hello.f90:6]: Yes, eventually

File: lib/spack/spack/build_systems/intel.py (hunk 4 above)
Comment (ax3l) [reply lib/spack/spack/build_systems/intel.py:1022]: > but users were trying to load the environment and build with other compilers than the one CUDA was built with.

I think the underlying problem is, that the compiler that cuda was installed with is irrelevant and could maybe be stripped from the spec. The cuda sdk installer just extracts binaries and and sources, as far as I am aware.
